packages/yarnpkg-core/sources/Project.ts

```diff
@@ -2015,6 +2061,196 @@ function applyVirtualResolutionMutations({
+
+        if (typeof resolvedDescriptor !== `undefined`) {
+          const peerResolution = getPackageFromDescriptor(resolvedDescriptor);
+          const peerVersion = peerResolution.version ?? `0.0.0`;
+
+          const ranges = new Set<string>();
+          for (const linkHash of linkHashes) {
+            const link = allPackages.get(linkHash);
+            if (typeof link === `undefined`)
+              throw new Error(`Assertion failed: Expected the link to be registered`);
+
+            const peerDependency = link.peerDependencies.get(ident.identHash);
+            if (typeof peerDependency === `undefined`)
+              throw new Error(`Assertion failed: Expected the ident to be registered`);
+
+            ranges.add(peerDependency.range);
+          }
+
+          const satisfiesAll = [...ranges].every(range => {
+            return semverUtils.satisfiesWithPrereleases(peerVersion, range);
+          });
+
+          if (!satisfiesAll) {
+            warnings.push({
+              type: WarningType.NotCompatible,
+              subject: dependent,
+              requested: ident,
+              requester: root,
+              version: peerVersion,
+              hash,
+            });
+          }
+        } else {
+          const peerDependencyMeta = root.peerDependenciesMeta.get(identStr);
+
+          if (!peerDependencyMeta?.optional) {
+            warnings.push({
+              type: WarningType.NotProvided,
+              subject: dependent,
+              requested: ident,
+              requester: root,
+              hash,
+            });
+          }
+        }
+      }
+    }
+  }
+
+  const warningSortCriterias: Array<((warning: Warning) => string)> = [
+    warning => structUtils.stringifyLocator(warning.subject),
+    warning => structUtils.stringifyIdent(warning.requested),
+    warning => `${warning.type}`,
+  ];
+
+  for (const warning of miscUtils.sortMap(warnings, warningSortCriterias)) {
+    switch (warning.type) {
+      case WarningType.NotProvided: {
+        report?.reportWarning(MessageName.MISSING_PEER_DEPENDENCY, `${structUtils.prettyLocator(project.configuration, warning.subject)} doesn't provide ${structUtils.prettyIdent(project.configuration, warning.requested)} (${formatUtils.pretty(project.configuration, warning.hash, formatUtils.Type.CODE)}), requested by ${structUtils.prettyIdent(project.configuration, warning.requester)}`);
+      } break;
+
+      case WarningType.NotCompatible: {
+        report?.reportWarning(MessageName.INCOMPATIBLE_PEER_DEPENDENCY, `${structUtils.prettyLocator(project.configuration, warning.subject)} provides ${structUtils.prettyIdent(project.configuration, warning.requested)} (${formatUtils.pretty(project.configuration, warning.hash, formatUtils.Type.CODE)}) with version ${structUtils.prettyReference(project.configuration, warning.version)}, which doesn't satisfy what ${structUtils.prettyIdent(project.configuration, warning.requester)} requests`);
```

Comment:
I really think we should mention the descendants if there are any because otherwise, the message might be technically incorrect, as the root requester might be satisfied, while the descendant requesters might not be satisfied.